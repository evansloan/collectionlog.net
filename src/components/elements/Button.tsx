import React from 'react';

type HTMLButtonProps = DetailedHTMLElementProps<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
>;

export enum ButtonType {
  STANDARD,
  CUTOUT
}

interface OverrideProps {
  title: string;
  icon?: string;
  iconAlt?: string;
  externalLink?: string;
  buttonType?: ButtonType;
}

type ButtonProps = HTMLButtonProps & OverrideProps;

const Button = (props: ButtonProps) => {
  let buttonClass = 'btn-standard border-black';
  if (props.buttonType == ButtonType.CUTOUT) {
    buttonClass = 'btn-cutout bg-grey hover:bg-dark-red border-accent hover:border-darker-red';
  }

  let buttonContent = (
    <div className='flex justify-around items-center'>
      {props.icon &&
        <img src={props.icon} alt={props.iconAlt} className='w-[40px]' />
      }
      {props.title}
    </div>
  );

  // Wrap the button in an anchor if we should be redirecting outwards
  if (props.externalLink) {
    buttonContent = (
      <a href={props.externalLink} className='no-underline text-white'>
        {buttonContent}
      </a>
    );
  }

  return (
    <button
      {...props}
      className={`p-2 ${buttonClass} border-2 text-white text-shadow ${props.className ?? ''}`}
    >
      {buttonContent}
    </button>
  );
};

export default Button;
