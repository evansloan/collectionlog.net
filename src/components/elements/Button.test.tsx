import renderer from 'react-test-renderer';
import Button from './Button';

import discordIcon from '../../assets/images/discord.png';
import githubIcon from '../../assets/images/github.png';

describe(`${Button.name} component`, () => {
  test.each(['Search', 'Show Options'])(
    'should render with title %p',
    (title: string) => {
      const render = renderer.create(<Button title={title}/>).toJSON();
      expect(render).toMatchSnapshot();
    });

  test.each([
    ['discord', discordIcon],
    ['github', githubIcon],
  ])(
    'should render with icon %p',
    (iconAltText: string, iconBase64: string) => {
      const render = renderer.create(<Button title='test' icon={iconBase64} iconAlt={iconAltText}/>).toJSON();
      expect(render).toMatchSnapshot();
    });

  test.each(['https://runelite.net/', 'https://example.com'])(
    'should render with external link %p',
    (externalLink: string) => {
      const render = renderer.create(<Button title='test' externalLink={externalLink}/>).toJSON();
      expect(render).toMatchSnapshot();
    }
  );
});

