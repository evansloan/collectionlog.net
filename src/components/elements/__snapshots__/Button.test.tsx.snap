// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render with external link "https://example.com" 1`] = `
<button
  className="p-2 btn-standard border-black border-2 text-white text-shadow "
>
  <a
    href="https://example.com"
  >
    <div
      className="flex justify-around items-center"
    >
      test
    </div>
  </a>
</button>
`;

exports[`Button component should render with external link "https://runelite.net/" 1`] = `
<button
  className="p-2 btn-standard border-black border-2 text-white text-shadow "
>
  <a
    href="https://runelite.net/"
  >
    <div
      className="flex justify-around items-center"
    >
      test
    </div>
  </a>
</button>
`;

exports[`Button component should render with icon "discord" 1`] = `
<button
  className="p-2 btn-standard border-black border-2 text-white text-shadow "
>
  <div
    className="flex justify-around items-center"
  >
    <img
      alt="discord"
      className="w-[40px]"
      src="discord.png"
    />
    test
  </div>
</button>
`;

exports[`Button component should render with icon "github" 1`] = `
<button
  className="p-2 btn-standard border-black border-2 text-white text-shadow "
>
  <div
    className="flex justify-around items-center"
  >
    <img
      alt="github"
      className="w-[40px]"
      src="github.png"
    />
    test
  </div>
</button>
`;

exports[`Button component should render with title "Search" 1`] = `
<button
  className="p-2 btn-standard border-black border-2 text-white text-shadow "
>
  <div
    className="flex justify-around items-center"
  >
    Search
  </div>
</button>
`;

exports[`Button component should render with title "Show Options" 1`] = `
<button
  className="p-2 btn-standard border-black border-2 text-white text-shadow "
>
  <div
    className="flex justify-around items-center"
  >
    Show Options
  </div>
</button>
`;
